src/storage/checkpoint/rle_sort.cpp

```diff
@@ -0,0 +1,323 @@
+		                    make_unique<BoundReferenceExpression>(keys_types[i], indexes[i]));
+	}
+
+	// create payload layout
+	payload_layout.Initialize(payload_types);
+
+	// initialize the global sort state
+	global_sort_state = make_unique<GlobalSortState>(buffer_manager, orders, payload_layout);
+}
+
+unique_ptr<FunctionData> RowGroupSortBindData::Copy() const {
+	return make_unique<RowGroupSortBindData>(payload_types, keys_types, indexes, db);
+}
+
+bool RowGroupSortBindData::Equals(const FunctionData &other) const {
+	return false;
+}
+
+RowGroupSortBindData::~RowGroupSortBindData() {
+}
+
+RLESort::RLESort(RowGroup &row_group, DataTable &data_table, TableDataWriter &writer,
+                 vector<ColumnCheckpointInfo> &infos)
+    : row_group(row_group), data_table(data_table), writer(writer), old_count(row_group.count) {
+	// Reorder columns to optimize RLE Compression - We skip if the table has indexes or is empty
+	if (!row_group.db.config.preserve_insertion_order && row_group.count != 0 && row_group.table_info.indexes.Empty()) {
+		// collect logical types by iterating the columns
+		for (idx_t column_idx = 0; column_idx < row_group.columns.size(); column_idx++) {
+			auto &column = row_group.columns[column_idx];
+			auto type_id = column->type.id();
+			auto column_compression = infos[column_idx].compression_type;
+			// We basically only sort columns with RLE compression and that are supported by the RLE algorithm
+			if (SupportedKeyType(type_id) && (column_compression == CompressionType::COMPRESSION_RLE)) {
+				// Gather types and ids of key columns (i.e., the ones we will sort on)
+				key_column_ids.push_back(column_idx);
+				key_column_types.push_back(column->type);
+			}
+			if (!SupportedPayloadType(type_id)) {
+				// We don't support RLE reordering on this table
+				key_column_ids.clear();
+				key_column_types.clear();
+				return;
+			}
+			// Gather types and ids of payload columns (i.e., the whole table)
+			payload_column_ids.push_back(column_idx);
+			payload_column_types.push_back(column->type);
+		}
+	}
+}
+
+bool RLESort::SupportedKeyType(LogicalTypeId type_id) {
+	switch (type_id) {
+	case LogicalTypeId::STRUCT:
+	case LogicalTypeId::LIST:
+	case LogicalTypeId::MAP:
+	case LogicalTypeId::TABLE:
+	case LogicalTypeId::ENUM:
+	case LogicalTypeId::AGGREGATE_STATE:
+	case LogicalTypeId::VARCHAR:
+	case LogicalTypeId::BLOB:
+	case LogicalTypeId::INTERVAL:
+	case LogicalTypeId::UUID:
+		return false;
+	default:
+		return true;
+	}
+}
+
+bool RLESort::SupportedPayloadType(LogicalTypeId type_id) {
+	switch (type_id) {
+	case LogicalTypeId::STRUCT:
+	case LogicalTypeId::MAP:
+	case LogicalTypeId::TABLE:
+	case LogicalTypeId::ENUM:
+	case LogicalTypeId::AGGREGATE_STATE:
+	case LogicalTypeId::INTERVAL:
+	case LogicalTypeId::UUID:
+		return false;
+	default:
+		return true;
+	}
+}
+
+bool RLESort::NewScoresBetter(RowGroup &sorted_rowgroup, vector<ColumnCheckpointInfo> &infos) {
```

Comment:
I have added the benefit analyser as well, which checks whether the sorted RowGroup has a lower score (i.e. smaller file size) than before. The RowGroup is only replaced when this is the case